src/linear_algebra/finsupp.lean

```diff
@@ -313,10 +313,41 @@ finsupp.sum_single_index (f i).map_zero
 
 theorem lsum_symm_apply (f : (α →₀ M) →ₗ[R] N) (x : α) : lsum.symm f x = f.comp (lsingle x) := rfl
 
+section
+variables (M) (R) (X : Type*)
+
+/--
+The type-level equivalence between R-linear maps from `R` to `M`, and points of `M` itself.
+This says that the forgetful functor from `R`-modules to types is representable, by `R`.
+-/
+@[simps]
+def ring_lmap_equiv_self : (R →ₗ[R] M) ≃ M :=
+{ to_fun := λ f, f 1,
+  inv_fun := λ m, { map_smul' := λ r r', by simp [mul_smul], ..(smul_add_hom R M).flip m },
+  left_inv := λ x, by { ext, simp },
+  right_inv := λ x, by { simp } }
+
+/--
+A slight rearrangement from `lsum` gives us
+the bijection underlying the free-forgetful adjunction for R-modules.
+-/
+noncomputable def hom_equiv : ((X →₀ R) →ₗ[R] M) ≃ (X → M) :=
+(lsum : _ ≃+ _).to_equiv.symm.trans $ equiv.arrow_congr (equiv.refl _) (ring_lmap_equiv_self M R)
```

Comment:
I'd be tempted to write this in reverse as
```suggestion
/--
A slight rearrangement from `lsum` gives us
the bijection underlying the free-forgetful adjunction for R-modules.
-/
noncomputable def lift :  (X → M) ≃ ((X →₀ R) →ₗ[R] M) :=
(equiv.arrow_congr (equiv.refl _) (ring_lmap_equiv_self M R).symm).trans (lsum : _ ≃+ _).to_equiv
```
so that it looks more like `free_monoid.lift` etc

Reply:
Done!